Mark `proc_macro_decls_static` as always used

jyn514 opened the pull request:
This would have avoided a bug in https://github.com/rust-lang/rust/pull/104860.

In practice this shouldn't matter since nothing uses the query other than the `dead_code` lint, but this isn't documented as an internal-only query so it seems nice for it to be accurate. I think for `dead_code` it doesn't matter because the relevant code is generated by `rustc_builtin_macros` and isn't linted.

I think @tmiasko or @bjorn3 would be a good reviewer?

r? @tmiasko

bjorn3 commented on compiler/rustc_passes/src/dead.rs:
```diff
@@ -623,6 +623,11 @@ fn create_and_seed_worklist<'tcx>(
         check_foreign_item(tcx, &mut worklist, id);
     }
 
+    if let Some(static_) = tcx.proc_macro_decls_static(()) {
```
Would adding`#[used]` when creating this static work?

petrochenkov replied:
+1, if it's possible we should rely on regular user-available attributes instead of hardcoding stuff in the compiler, both for dead code and symbol export analysis.

tmiasko replied:
Interaction between dead code lint and #[used] attribute is tested in issue-41628.rs.

jyn514 committed:
Mark `proc_macro_decls_static` as always used

This would have avoided a bug in https://github.com/rust-lang/rust/pull/104860.

In practice this shouldn't matter since nothing uses the query other than the `dead_code` lint,
but this isn't documented as an internal-only query so it seems nice for it to be accurate.
I think for `dead_code` it doesn't matter because the relevant code is generated by `rustc_builtin_macros` and isn't linted.

tmiasko commented:
@bors r+

bors commented:
:pushpin: Commit 164e22109b697ed0bc7dd62039362b2fb5482ef9 has been approved by `tmiasko`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=tmiasko 164e22109b697ed0bc7dd62039362b2fb5482ef9 -->
<!-- homu: {"type":"Approved","sha":"164e22109b697ed0bc7dd62039362b2fb5482ef9","approver":"tmiasko","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
